### PR TITLE
[model, ci] feat: migrate qwen2_5_vl to transformers v5

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Followup of https://github.com/ByteDance-Seed/VeOmni/pull/652

> Migrates `qwen2_5vl` from the v4 monkey-patch path to the transformers v5 patchgen + generated modeling path. Produces `patched_modeling_qwen2_5_vl_gpu.py` via `make patchgen` from a new `qwen2_5_vl_gpu_patch_gen_config.py`, updates `__init__.py` to wire up the v5 path, and extends existing test suites to cover qwen2_5_vl under v5.

### Checklist Before Starting

- Search for relative PRs/issues and link here: depends on #652 (`/veomni-migrate-transformers-v5` skill)
- PR title follows `[{modules}] {type}: {description}` format

### Test

> - `tests/models/test_models_patch.py`, `tests/distributed/test_dummy_forward.py`, `tests/e2e/test_e2e_parallel.py`, `tests/models/test_vlm_trainer.py` extended with qwen2_5_vl v5 cases.
> - Verified `make patchgen` regenerates `generated/patched_modeling_qwen2_5_vl_gpu.py` and `.diff` deterministically.

### API and Usage Example

> No user-facing API changes. qwen2_5_vl continues to load via the existing VeOmni entry points; under transformers v5 the generated modeling is used automatically.

### Design & Code Changes

> - Add `veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py` with patchgen decorators (Ulysses SP in attention/forward, MoE-free dense VLM path, multimodal position-id handling).
> - Regenerate `generated/patched_modeling_qwen2_5_vl_gpu.{py,diff}` via `make patchgen` — do not hand-edit.
> - Update `veomni/models/transformers/qwen2_5vl/__init__.py` to select the v5 generated module when `transformers >= v5`, falling back to the v4 monkey patch otherwise.
> - Refine the `/veomni-migrate-transformers-v5` skill based on lessons learned during this migration.

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [ ] Applied pre-commit checks
- [ ] Added/updated documentation
- [ ] Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+# ================================================================
+@config.override_method(
+    "Qwen2_5_VLForConditionalGeneration.get_position_id_func",
+    description="Use VeOmni precomputed position-id function and unified multimodal token ids",
+)
+def qwen2_5_vl_get_position_id_func_patched(self):
+    def get_position_id(main_func, self, **kwargs):
+        position_ids, rope_deltas = main_func(self, **kwargs)  # position_ids (dim, bs, l)
+        return {"position_ids": position_ids, "rope_deltas": rope_deltas}
+
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py]: done

**Comment by TimYangst** [comment]
@gemini-code-assist please review the pr again.

File: veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+        inputs_embeds = gather_seq_scatter_heads(
+            inputs_embeds, seq_dim=1, head_dim=2, group=get_parallel_state().sp_group
+        )
+    # --- Patch.1 ---
+
+    if pixel_values is not None:
+        # --- Patch.6 ---
+        # v5 get_image_features returns BaseModelOutputWithPooling whose
+        # pooler_output is tuple[per-image tensor] after torch.split
+        image_embeds = self.get_image_features(pixel_values, image_grid_thw, return_dict=True).pooler_output
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py:461]: Not actual issues: pixel_values is replicated across SP ranks (only the token sequence is gathered inside
 the ViT), so grid_thw is identical on every rank and doesn't need gathering.

File: veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+            video_grid_thw=video_grid_thw,
+            second_per_grid_ts=second_per_grid_ts,
+            inputs_embeds=inputs_embeds,
+            attention_mask=attention_mask,
+            past_key_values=past_key_values,
+        )
+    else:
+        # --- Patch.5 ---
+        if position_ids.dim() == 3 and position_ids.shape[1] == 3:
+            position_ids = position_ids.transpose(0, 1).contiguous()  # bs, 3, l -> 3, bs, l
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py:525]: Not actual issues: It's emitted at module scope via config.add_post_import_block(...), so the generated file has it as a top-level function — picklable.

File: veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+
+    # --- Patch.2 ---
+    # Precomputed image/video masks when provided by VeOmni data pipeline,
+    # otherwise all-gather input_ids across SP group to compute them.
+    image_mask = kwargs.pop("image_mask", None)
+    video_mask = kwargs.pop("video_mask", None)
+    if video_mask is None and image_mask is None:
+        input_ids_list = [torch.zeros_like(input_ids) for _ in range(get_parallel_state().sp_size)]
+        dist.all_gather(input_ids_list, input_ids, group=get_parallel_state().sp_group)
+        image_mask, video_mask = self.get_placeholder_mask(torch.cat(input_ids_list, dim=0))
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen2_5vl/qwen2_5_vl_gpu_patch_gen_config.py:439]: Intentionally preserving v4 behavior in this migration PR:
  - Vision SP + global-attention flow (gather → window permute → scatter → blocks with full cu_seqlens): identical to the v4 monkey-patch, which runs in production.                                                                                                                          
  - torch.cat(input_ids_list, dim=0) and position_ids.shape[1] == 3 check: same as v4. If these are real bugs, they'd need a separate PR that fixes both v4 and v5 paths — out of scope for this migration.

**Review by FoolPlayer — APPROVED** [review]
